### PR TITLE
sda sipnet working with simulated data

### PR DESCRIPTION
**Pull request by araiho**
@mdietze 

sda.enkf works with the setup_da.R code for SIPNET. Moving on to tree ring code.

### CODE REVIEW

File: models/sipnet/R/read.restart.SIPNET.R [hunk 1]
```diff
@@ -1,29 +1,53 @@
...
-    #forecast$microbe[i] = NA
-  }
+  #Read ensemble output
+  ens <- read.output(runid = runid,outdir = file.path(outdir, runid),
+                     start.year = time, end.year=time,
+                     variables=variables)
+    
+    last = length(ens$NPP)
+  
+    forecast[[1]] <- mean(ens$NPP)*unit.conv ## kg C m-2 s-1 -> Mg/ha/yr [Check]
```
Comment (mdietze) [inline models/sipnet/R/read.restart.SIPNET.R]: if forecast is just holding scalars, why is it a list rather than a vector, since you just end up unlisting it a few lines later?

File: models/sipnet/R/write.restart.SIPNET.R [hunk 2]
```diff
@@ -1,8 +1,37 @@
...
+  if(RENAME == TRUE) {
+    file.rename(file.path(outdir,runid,"sipnet.out"),
+                file.path(outdir,runid,paste0("sipnet.out",time)))
+    system(paste("rm",file.path(rundir,runid,"sipnet.clim")))
+  }else{
+    print(paste("Files not renamed -- Need to rerun year",time,"before next time step"))
   }
+   #file.remove(file.path(settings$rundir,runid,"sipnet.clim"))
+  
+  settings$run$start.date <- paste0(time + 1,"/01/01")
```
Comment (mdietze) [inline models/sipnet/R/write.restart.SIPNET.R]: this seems really fragile.  As we generalize we'll need to be passing the new end time as the argument (and assuming that the previous end time is the new start time.

Comment (araiho) [inline models/sipnet/R/write.restart.SIPNET.R]: I thought that since we talked about it always being a yearly time step this would be okay.

File: modules/assim.sequential/R/sda.enkf.R [hunk 3]
```diff
@@ -3,49 +3,56 @@
...
 ##' 
 ##' @param settings    PEcAn settings object
-##' @param IC          data.frame of initial condition sample (nens X nstate)
-##' @param prior       data.frame of model parameter sample (nens X nstate)
 ##' @param obs.mean    data.frame of observations of the mean of variables (time X nstate)
 ##' @param obs.cov      data.frame of observations of the sd of variables (time X nstate)
-##' @param processvar  flag for if process variance should be estimated or not
-##' @param sample.parameters flag for if parameters should come from meta.analysis or be defaults. if false parameters are set to defaults
+##' @param pick.trait.params  character vector of traits to sample if sample.parameters == TRUE
+##' @param given.process.variance flag for if parameters should come from meta.analysis or be defaults. if false parameters are set to defaults
```
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: two things: 
1) why aren't these two in settings?
2) 'given.process.variance' is a very unintuitive name for a flag about the meta.analysis.

File: models/sipnet/R/write.restart.SIPNET.R [hunk 4]
```diff
@@ -1,8 +1,37 @@
...
+  settings$run$end.date <- paste0(time + 1,"/12/31")
+    
+  if(sample_parameters == TRUE){
+    load(file.path(settings$outdir, "samples.Rdata"))
+    do.call(my.write.config, args = list(defaults = NULL, trait.values = trait.values,
```
Comment (araiho) [inline models/sipnet/R/write.restart.SIPNET.R]: I thought samples.Rdata were the original samples.

my.write.restart takes my.write.config

File: modules/assim.sequential/R/sda.enkf.R [hunk 5]
```diff
@@ -3,49 +3,56 @@
...
-  
-  #   if(nrow(prior) == 1 | is.null(nrow(prior))){
-  #     var.names = names(prior)
-  #     prior = as.data.frame(matrix(rep(prior,each=nens),nrow=nens))
-  #     names(prior) = var.names
-  #   }
+  forecast.time.step <- settings$state.data.assimilation$forecast.time.step
+  spin.up.start <- strftime(settings$state.data.assimilation$spin.up$start.date,"%Y")
+  spin.up.end <- strftime(settings$state.data.assimilation$spin.up$end.date,"%Y")
+  nens = settings$ensemble$size#settings$state.data.assimilation$n.ensemble #which one?
```
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: It's be better to get the ensemble size from settings$state.data.assimilation$n.ensemble because there's nothing that says the ensemble analysis and SDA should have the same ensemble size, plus specifying ensemble$size has the unwanted consequence of turning on the EA in the workflow (not what we want)

File: modules/assim.sequential/R/sda.enkf.R (hunk 3 above)
Comment (araiho) [inline modules/assim.sequential/R/sda.enkf.R]: 1. because for linkages you can't do all of the parameters. you have to pick a subset. it is in the settings but not in the sda part in the set up. I can move it over.
2. given.process.variance is supposed to be from DA that you did before not from the meta analysis.

File: modules/assim.sequential/R/sda.enkf.R [hunk 6]
```diff
@@ -73,7 +84,9 @@ sda.enkf <- function(settings, IC, prior, obs.mean, obs.cov, variables,
     ensemble.id <- -1
   }
   
-  ## model-specific functions
+  ###-------------------------------------------------------------------###
+  ### get model specific functions                                      ###
+  ###-------------------------------------------------------------------### 
   do.call("require",list(paste0("PEcAn.",model)))
   my.write.config <- paste("write.config.",model,sep="")
```
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: you also need a my.write.config

Comment (araiho) [inline modules/assim.sequential/R/sda.enkf.R]: What do you mean? There is a my.write.config?

File: modules/assim.sequential/R/sda.enkf.R [hunk 7]
```diff
@@ -83,19 +96,11 @@ sda.enkf <- function(settings, IC, prior, obs.mean, obs.cov, variables,
...
-  #   file.copy(full.met,new.met)
-  #   met <- new.met#split.met.SIPNET(new.met)
-  
-  
+ 
   ###-------------------------------------------------------------------###
   ### perform initial set of runs                                       ###
   ###-------------------------------------------------------------------###  
-  X = IC
+  X = matrix(NA,as.numeric(settings$ensemble$size),length(settings$pft)) #why does X need to be here?
```
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: You've dropped the IC as an argument to the function without moving it into the settings or moving the IC generating function into sda.enkf. Then you set the IC matrix to all NA's, so I'm not sure how SIPNET is going to run if you give it NA for all its initial biomass pools.  Also, the dimension should be ensembles by state variables, not ensembles by PFTs

File: modules/assim.sequential/R/sda.enkf.R [hunk 8]
```diff
@@ -121,16 +126,20 @@ sda.enkf <- function(settings, IC, prior, obs.mean, obs.cov, variables,
...
-                                           settings = settings, run.id = run.id[[i]]))
+      do.call(my.write.config, args = list(defaults = NULL, trait.values = lapply(ensemble.samples, function(x, n){x[i,]},n = i),
+                                           settings = settings, run.id = run.id[[i]],
+                                           inputs = list(met=list(path=met[1]))))
     } else {
-      do.call(my.write.config,args=list(trait.values = NA,settings=settings,run.id = run.id[[i]],restart=FALSE))
+      load(file.path(settings$outdir, paste0("ensemble.samples.",settings$state.data.assimilation$prior,".Rdata")))
+      do.call(my.write.config,args=list(defaults = NULL, trait.values = ens.samples, 
+                                        settings=settings,run.id = run.id[[i]],restart=FALSE,
+                                        inputs = list(met=list(path=met[1])),IC=IC[i,]))
```
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: where is IC defined?

File: modules/assim.sequential/R/sda.enkf.R [hunk 9]
```diff
@@ -240,42 +255,58 @@ sda.enkf <- function(settings, IC, prior, obs.mean, obs.cov, variables,
...
+    ###-------------------------------------------------------------------###  
+    if(any(obs)){ #if no observations skip analysis
     mu.f = as.numeric(apply(X,2,mean,na.rm=TRUE))
     Pf   = cov(X)
-    Y    = obs.mean[[t]][[1]][pmatch(colnames(X), names(obs.mean[[t]][[1]]))]
+    Y    = obs.mean[[t]][pmatch(colnames(X), names(obs.mean[[t]]))]
 
-    H = diag(length(obs.mean[[t]][[1]]))
-    R = obs.cov[[t]]#diag(as.numeric(obs.sd[[t]][[1]][pmatch(colnames(X),names(obs.mean[[t]][[1]]))])^2)
+    R = as.matrix(obs.cov[[t]])#how do we make sure this is in the right order?
```
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: Doesn't it have to be in the same order as obs.mean?  That seems like a safe assumption to me

File: modules/assim.sequential/R/sda.enkf.R [hunk 10]
```diff
@@ -240,42 +255,58 @@ sda.enkf <- function(settings, IC, prior, obs.mean, obs.cov, variables,
...
+        }
       }
     }
-   
+
+    ###-------------------------------------------------------------------###
+    ### Kalman Filter                                                     ###
+    ###-------------------------------------------------------------------###
     if(processvar == FALSE){
+      H = diag(ncol(X)) #correct fix?
```
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: unfortunately no. That only works if X and Y are the same size and match 1:1, which is going to be pretty rare and doesn't apply to the SIPNET example we need to get working

File: modules/assim.sequential/R/sda.enkf.R (hunk 7 above)
Comment (araiho) [inline modules/assim.sequential/R/sda.enkf.R]: I think that X is unnecessary because it gets created as a list in read.restart later. 

Right now, SIPNET is running off its default pools and the initial conditions are varied by sampling over all of the parameters during spin up. 

I'm not sure how to generalize this because linkages and other models don't use IC.

File: modules/assim.sequential/R/sda.enkf.R (hunk 8 above)
Comment (araiho) [inline modules/assim.sequential/R/sda.enkf.R]: Sorry, meant to remove that. It is not defined per explanation above.

File: modules/assim.sequential/R/sda.enkf.R [hunk 11]
```diff
@@ -240,42 +255,58 @@ sda.enkf <- function(settings, IC, prior, obs.mean, obs.cov, variables,
...
+      ###-------------------------------------------------------------------###
+      
       #### initial conditions
       bqq[1] <- length(mu.f)
       aqq[1,,] <- diag(length(mu.f))*bqq[1]
       
-      #### extinct in all ensemble vector
-      E <- which(colSums(X)==0)
+      #### extinct from the forecast ensembles
+      E <- which(colSums(Pf) == 0)
```
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: This needs to be generalized -- the issue isn't just species becoming extinct (in the general case, there's no guaranteed that either  X or Y contain state variables about species abundance).

File: modules/assim.sequential/R/sda.enkf.R [hunk 12]
```diff
@@ -240,42 +255,58 @@ sda.enkf <- function(settings, IC, prior, obs.mean, obs.cov, variables,
...
       F2M <- na.omit(F2M)
       
-      #### translating x to y so remove extinct species
+      #### translating x to y ## remove X that you don't have data for
       X2Y<- seq(1,length(Y),1)
-      X2Y <- X2Y[!is.na(Y)]
+      X2Y <- X2Y[which(!is.na(Y))]
+      
+      for(i in 1:length(diag(Pf))){
+        if(diag(Pf)[i]<.0000001) diag(Pf)[i]<-1 #HACK
```
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: needs a better fix

File: modules/assim.sequential/R/sda.enkf.R [hunk 13]
```diff
@@ -329,71 +360,92 @@ sda.enkf <- function(settings, IC, prior, obs.mean, obs.cov, variables,
...
                 args=list(out.dir = outdir, runid = run.id[[i]],
                           time = total.time[t], settings = settings,
                           analysis.vec = analysis[i,],
                           RENAME = TRUE, PLOT=FALSE, variables=variables,
                           sample_parameters = sample_parameters,
-                          trait.values = lapply(ensemble.samples,
-                                                function(x, n){x[n,pick.trait.params]},
-                                                n = i)))
+                          trait.values = trait.values,
+                          my.write.config,met=met))
```
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: drop my.write.config

File: modules/assim.sequential/R/sda.enkf.R [hunk 14]
```diff
@@ -329,71 +360,92 @@ sda.enkf <- function(settings, IC, prior, obs.mean, obs.cov, variables,
...
-  load(climate_file)
-  temp.mat <- temp.mat[total.time-853,]
-  precip.mat <- precip.mat[total.time-853,]
   
+  ###-------------------------------------------------------------------###
+  ### create diagnostics                                                ###
+  ###-------------------------------------------------------------------### 
+  
+  ### LOAD CLIMATE ### HACK ### LINKAGES SPECIFIC
+  # climate_file <- settings$run$inputs$met$path
```
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: at least embed this hack in an if(model == "LINKAGES) or something like that

File: models/sipnet/R/write.restart.SIPNET.R (hunk 2 above)
Comment (mdietze) [inline models/sipnet/R/write.restart.SIPNET.R]: it's OK for the moment, but it would be good to keep the yearly timestep assumption in EnKF, not build it into the model-specific functions.

File: models/sipnet/R/write.restart.SIPNET.R (hunk 4 above)
Comment (mdietze) [inline models/sipnet/R/write.restart.SIPNET.R]: parameter sampling should be isolated to the start of the ensemble, and then ensemble members should carry their parameters forward with them, not read them from some general file.  That said, when you do sample parameters you need to make sure to get the correct posterior ID's from the settings file. People will be running SDA after PDA and we want them to be using those posteriors.

File: modules/assim.sequential/R/sda.enkf.R (hunk 3 above)
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: I understand why the first variable exists, I don't understand why you didn't put it in the settings object

Don't following the second point, could you rephrase?

File: modules/assim.sequential/R/sda.enkf.R (hunk 11 above)
Comment (araiho) [inline modules/assim.sequential/R/sda.enkf.R]: extinct is probably the wrong name but this works in two cases 1) the model gives you the same output for every ensemble therefore cov is 0 2) the model gives you no output (zeros) therefore the cov is 0. I'm not sure what else to do to get around this?

File: modules/assim.sequential/R/sda.enkf.R (hunk 6 above)
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: meant to write my.write.restart, which seems to exist below but I didn't see where it was defined.

File: modules/assim.sequential/R/sda.enkf.R (hunk 7 above)
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: For now LINKAGES can just ignore that argument, but it IC do have to make it into SIPNET.

All models **DO** have IC in some form

File: models/sipnet/R/write.restart.SIPNET.R (hunk 4 above)
Comment (araiho) [inline models/sipnet/R/write.restart.SIPNET.R]: Okay sounds good. There's a place for it in the state.data.assimilation xml already and would happen now if you selected FALSE for sample_parameters and gave the posterior value you wanted.

But, I have a questions about sample.parameters(). sample.parameters created samples.R not anything else with posteriors. And the ensemble.samples.posterior##.Rdata is just one set. Which function creates multiple ensembles of parameters in posterior files like samples.R?

File: modules/assim.sequential/R/sda.enkf.R (hunk 3 above)
Comment (araiho) [inline modules/assim.sequential/R/sda.enkf.R]: I thought that we were going to assume that you got your process variance from another analysis (maybe outside of pecan) if you were going to define given.process.variance.

File: modules/assim.sequential/R/sda.enkf.R (hunk 6 above)
Comment (araiho) [inline modules/assim.sequential/R/sda.enkf.R]: got it

File: modules/assim.sequential/R/sda.enkf.R (hunk 7 above)
Comment (araiho) [inline modules/assim.sequential/R/sda.enkf.R]: Right, of some form. I think we have to decide what form before I can figure out what to put that will be general.

File: modules/assim.sequential/R/sda.enkf.R (hunk 11 above)
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: Ok, so just clean up comments. The latter (all 0s) just seems like a special case of no variance -- should be handled fine by the general case

File: modules/assim.sequential/R/sda.enkf.R (hunk 12 above)
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: I understand, but it needs a better fix. In SIPNET all ensemble members will have zero snow midsummer, but that doesn't make them wrong and it surely doesn't mean we have a 1m SD uncertainty about midsummer snow. Just an example

Comment (araiho) [inline modules/assim.sequential/R/sda.enkf.R]: Oh I see. Well, the reason it's a problem is because solve won't work because it's base tolerance is 2.220446e-16. You can set it to 0 but then JAGS gets upset. So, maybe I'll just set it to the default tolerance for solve?

File: modules/assim.sequential/R/sda.enkf.R (hunk 3 above)
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: Wait, are you saying "given.process.variance" is the covariance matrix Q?  That's a really unintuitive name and bears no relation to what you wrote in the documentation, so I didn't understand what you were describing.

Also, I can only find the variable used once in the code, and there it's used incorrectly 
```
 Pa = given.process.variance 
```
If there's no data and no process error then Pa = Pf (in the existing code there seems to be a typo that says Pa = Pa). If there's no data and process error then Pa = Pf + Q.  Actually, the latter applies even when there is data -- you need to update the standard EnKF for the case where process error is given as well.

File: modules/assim.sequential/R/sda.enkf.R (hunk 7 above)
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: Start by reverting to the previous workflow -- for now take IC as a function argument and use the existing sipnet IC function to generate that IC. When using LINKAGES you can default IC to NULL since you don't use it

File: modules/assim.sequential/R/sda.enkf.R (hunk 12 above)
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: No, having a 0 on the diagonal is going to cause bad things to happen -- I thought the code above where you solve for E already identified this case an excluded it?  

Hadn't we discussed setting it to the minimally observed variance rather than an arbitrary constant?  We need to think about this more, but for now that seems like a better fix than hard-coding 1.

File: modules/assim.sequential/R/sda.enkf.R [hunk 15]
```diff
@@ -100,7 +100,6 @@ sda.enkf <- function(settings, obs.mean, obs.cov, pick.trait.params = NULL,
   ###-------------------------------------------------------------------###
   ### perform initial set of runs                                       ###
   ###-------------------------------------------------------------------###  
-  X = matrix(NA,as.numeric(settings$ensemble$size),length(settings$pft)) #why does X need to be here?
```
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: Needs to go back to X = IC

File: modules/assim.sequential/R/sda.enkf.R [hunk 16]
```diff
@@ -261,14 +260,14 @@ sda.enkf <- function(settings, obs.mean, obs.cov, pick.trait.params = NULL,
     if(any(obs)){ #if no observations skip analysis
     mu.f = as.numeric(apply(X,2,mean,na.rm=TRUE))
     Pf   = cov(X)
-    Y    = obs.mean[[t]][pmatch(colnames(X), names(obs.mean[[t]]))]
+    Y    = na.omit(obs.mean[[t]][pmatch(colnames(X), names(obs.mean[[t]]))])
```
Comment (mdietze) [inline modules/assim.sequential/R/sda.enkf.R]: This won't work as coded. If you do a na.omit on Y, then Y and R won't line up anymore. You need to detect the NA's in Y and remove those rows & columns from R as well.  You should also throw a warning because Y shouldn't have NAs in it.

File: modules/assim.sequential/R/sda.enkf.R (hunk 3 above)
Comment (araiho) [inline modules/assim.sequential/R/sda.enkf.R]: So we want to estimate Q even if there isn't data?
